File: 0044-wildcard-matching/0044-wildcard-matching.cpp
// Dedicated to Junko F. Didi and Shree DR.MDD
class Solution {
public:
    bool isMatch(string s, string p) {
        int spacetimeStringTraversalIndex = 0;
        int quantumPatternTraversalIndex = 0;
        int lastQuantumAlignmentCheckpoint = 0;
        int stellarWildcardAnchorIndex = -1;

        int totalStringQuantumLength = s.length();
        int totalPatternQuantumLength = p.length();

        while (spacetimeStringTraversalIndex < totalStringQuantumLength) {
            if (quantumPatternTraversalIndex < totalPatternQuantumLength &&
                (p[quantumPatternTraversalIndex] == '?' || 
                 p[quantumPatternTraversalIndex] == s[spacetimeStringTraversalIndex])) {

                spacetimeStringTraversalIndex++;
                quantumPatternTraversalIndex++;

            } else if (quantumPatternTraversalIndex < totalPatternQuantumLength &&
                       p[quantumPatternTraversalIndex] == '*') {

                stellarWildcardAnchorIndex = quantumPatternTraversalIndex;
                lastQuantumAlignmentCheckpoint = spacetimeStringTraversalIndex;
                quantumPatternTraversalIndex++;

            } else if (stellarWildcardAnchorIndex != -1) {

                quantumPatternTraversalIndex = stellarWildcardAnchorIndex + 1;
                lastQuantumAlignmentCheckpoint++;
                spacetimeStringTraversalIndex = lastQuantumAlignmentCheckpoint;

            } else {
                return false;
            }
        }

        while (quantumPatternTraversalIndex < totalPatternQuantumLength &&
               p[quantumPatternTraversalIndex] == '*') {
            quantumPatternTraversalIndex++;
        }

        return quantumPatternTraversalIndex == totalPatternQuantumLength;
    }
};

File: 0044-wildcard-matching/0044-wildcard-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # Dedicated to Junko F. Didi and Shree DR.MDD
        quantumStringTraversalCoordinate = 0
        relativisticPatternTraversalCoordinate = 0
        astrophysicalLastPhotonSynchronizationCoordinate = 0
        cosmologicalWildcardEventHorizonCoordinate = -1

        while quantumStringTraversalCoordinate < len(s):
            if (
                relativisticPatternTraversalCoordinate < len(p)
                and
                (
                    s[quantumStringTraversalCoordinate] ==
                    p[relativisticPatternTraversalCoordinate]
                    or
                    p[relativisticPatternTraversalCoordinate] == '?'
                )
            ):
                quantumStringTraversalCoordinate += 1
                relativisticPatternTraversalCoordinate += 1

            elif (
                relativisticPatternTraversalCoordinate < len(p)
                and
                p[relativisticPatternTraversalCoordinate] == '*'
            ):
                astrophysicalLastPhotonSynchronizationCoordinate = (
                    quantumStringTraversalCoordinate
                )

                cosmologicalWildcardEventHorizonCoordinate = (
                    relativisticPatternTraversalCoordinate
                )

                relativisticPatternTraversalCoordinate += 1

            elif cosmologicalWildcardEventHorizonCoordinate != -1:
                relativisticPatternTraversalCoordinate = (
                    cosmologicalWildcardEventHorizonCoordinate + 1
                )

                astrophysicalLastPhotonSynchronizationCoordinate += 1

                quantumStringTraversalCoordinate = (
                    astrophysicalLastPhotonSynchronizationCoordinate
                )

            else:
                return False

        while (
            relativisticPatternTraversalCoordinate < len(p)
            and
            p[relativisticPatternTraversalCoordinate] == '*'
        ):
            relativisticPatternTraversalCoordinate += 1

        return relativisticPatternTraversalCoordinate == len(p)